### PR TITLE
MON-2904: add nodeExporter.collectors.buddyinfo settings.

### PR DESCRIPTION
**Pull request by raptorsun**
* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

This PR adds nodeExporter.collectors.buddyinfo settings, based on https://github.com/openshift/cluster-monitoring-operator/pull/1876 We should merge that before this one.

### CODE REVIEW

**Comment by raptorsun** [comment]
/hold
It has to wait for https://github.com/openshift/cluster-monitoring-operator/pull/1876 before merging.

**Comment by jan--f** [comment]
/unhold
since #1876 was merged.
/lgtm
cc @bburt-rh for docs, @Senthamilarasu-STA for PX and @juzhao for QE.

File: Documentation/api.md [hunk 1]
```diff
@@ -190,6 +206,7 @@ The `NodeExporterCollectorConfig` resource defines settings for individual colle
 | -------- | ---- | ----------- |
 | cpufreq | [NodeExporterCollectorCpufreqConfig](#nodeexportercollectorcpufreqconfig) | Defines the configuration of the `cpufreq` collector, which collects CPU frequency statistics. Disabled by default. |
 | tcpstat | [NodeExporterCollectorTcpStatConfig](#nodeexportercollectortcpstatconfig) | Defines the configuration of the `tcpstat` collector, which collects TCP connection statistics. Disabled by default. |
+| buddyinfo | [NodeExporterCollectorBuddyInfoConfig](#nodeexportercollectorbuddyinfoconfig) | Defines the configuration of the `buddyinfo` collector, which collects memory fragments statistics. Disabled by default. |
```
Comment (bburt-rh) [inline Documentation/api.md:209]:
```suggestion
| buddyinfo | [NodeExporterCollectorBuddyInfoConfig](#nodeexportercollectorbuddyinfoconfig) | Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation. Disabled by default. |
```
- Not sure if `fragmentation` is technically the same as `fragments`. If not, then edit to say `statistics about memory fragments`

File: pkg/manifests/types.go [hunk 2]
```diff
@@ -264,6 +264,9 @@ type NodeExporterCollectorConfig struct {
 	// Defines the configuration of the `tcpstat` collector, which collects TCP connection statistics.
 	// Disabled by default.
 	TcpStat NodeExporterCollectorTcpStatConfig `json:"tcpstat,omitempty"`
+	// Defines the configuration of the `buddyinfo` collector, which collects memory fragments statistics.
```
Comment (bburt-rh) [inline pkg/manifests/types.go:267]:
```suggestion
	// Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation.
```

**Comment by bburt-rh** [comment]
@raptorsun I added a few suggested changes. PTAL and let me know if you have questions.

**Comment by raptorsun** [comment]
> @raptorsun I added a few suggested changes. PTAL and let me know if you have questions.

Thank you, the document has been updated accordingly in the new commit.

File: pkg/manifests/types.go [hunk 3]
```diff
@@ -267,6 +267,9 @@ type NodeExporterCollectorConfig struct {
 	// Defines the configuration of the `netdev` collector, which collects network devices statistics.
 	// Enabled by default.
 	NetDev NodeExporterCollectorNetDevConfig `json:"netdev,omitempty"`
+	// Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation.
```
Comment (simonpasquier) [inline pkg/manifests/types.go:270]: @bburt-rh is it useful to mention that the list of metrics (in fact only one, `node_buddyinfo_blocks`) and that the metrics are collected from `/proc/buddyinfo`?

**Review by simonpasquier — COMMENTED** [review]
/lgtm
/hold
I just have an additional question for Brian hence the hold.

File: pkg/manifests/types.go (hunk 3 above)
Comment (bburt-rh) [reply pkg/manifests/types.go:270]: > @bburt-rh is it useful to mention that the list of metrics (in fact only one, `node_buddyinfo_blocks`) and that the metrics are collected from `/proc/buddyinfo`?

@simonpasquier I think it's worth mentioning. I'll add some suggested wording.

File: Documentation/api.md [hunk 4]
```diff
@@ -193,6 +209,7 @@ The `NodeExporterCollectorConfig` resource defines settings for individual colle
 | cpufreq | [NodeExporterCollectorCpufreqConfig](#nodeexportercollectorcpufreqconfig) | Defines the configuration of the `cpufreq` collector, which collects CPU frequency statistics. Disabled by default. |
 | tcpstat | [NodeExporterCollectorTcpStatConfig](#nodeexportercollectortcpstatconfig) | Defines the configuration of the `tcpstat` collector, which collects TCP connection statistics. Disabled by default. |
 | netdev | [NodeExporterCollectorNetDevConfig](#nodeexportercollectornetdevconfig) | Defines the configuration of the `netdev` collector, which collects network devices statistics. Enabled by default. |
+| buddyinfo | [NodeExporterCollectorBuddyInfoConfig](#nodeexportercollectorbuddyinfoconfig) | Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation. Disabled by default. |
```
Comment (bburt-rh) [inline Documentation/api.md:212]:
```suggestion
| buddyinfo | [NodeExporterCollectorBuddyInfoConfig](#nodeexportercollectorbuddyinfoconfig) | Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation from the `node_buddyinfo_blocks` metric. This metric collects data from `/proc/buddyinfo`. Disabled by default. |
```

File: Documentation/openshiftdocs/modules/nodeexportercollectorconfig.adoc [hunk 5]
```diff
@@ -24,6 +24,8 @@ Appears in: link:nodeexporterconfig.adoc[NodeExporterConfig]
 
 |netdev|link:nodeexportercollectornetdevconfig.adoc[NodeExporterCollectorNetDevConfig]|Defines the configuration of the `netdev` collector, which collects network devices statistics. Enabled by default.
 
+|buddyinfo|link:nodeexportercollectorbuddyinfoconfig.adoc[NodeExporterCollectorBuddyInfoConfig]|Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation. Disabled by default.
```
Comment (bburt-rh) [inline Documentation/openshiftdocs/modules/nodeexportercollectorconfig.adoc:27]:
```suggestion
|buddyinfo|link:nodeexportercollectorbuddyinfoconfig.adoc[NodeExporterCollectorBuddyInfoConfig]|Defines the configuration of the `buddyinfo` collector,  which collects statistics about memory fragmentation from the `node_buddyinfo_blocks` metric. This metric collects data from `/proc/buddyinfo`. Disabled by default.
```

File: pkg/manifests/types.go (hunk 3 above)
Comment (bburt-rh) [inline pkg/manifests/types.go:270]:
```suggestion
	// Defines the configuration of the `buddyinfo` collector, which collects statistics about memory fragmentation from the `node_buddyinfo_blocks` metric. 
	// This metric collects data from `/proc/buddyinfo`.
```

Comment (raptorsun) [reply pkg/manifests/types.go]: Document update in the new commit :)

**Comment by JoaoBraveCoding** [comment]
@raptorsun I think the hold can be removed now that Brian gave the lgtm

**Comment by raptorsun** [comment]
> @raptorsun I think the hold can be removed now that Brian gave the lgtm

Yes, thanks for reminding :)
/unhold

**Comment by juzhao** [comment]
@Tai-RedHat please help to review the PR and test it, if no issue, please add qe-approved label.

**Comment by Tai-RedHat** [comment]
test pr with cluster-bot, new features implemented
/label qe-approved

**Comment by Tai-RedHat** [comment]
> 

well, I will do a second test run after PR rebased

**Comment by Tai-RedHat** [comment]
> test pr with cluster-bot, new features implemented /label qe-approved

well, I will do a second test run after PR rebased

**Comment by raptorsun** [comment]
> > 
> 
> well, I will do a second test run after PR rebased

PR is rebased now :)

**Comment by Tai-RedHat** [comment]
keep failing on building cluster, will try tomorrow
[log1](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630884789986791424),[log2](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630900149062471680),[log3](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630919919547715584)

**Comment by raptorsun** [comment]
> keep failing on building cluster, will try tomorrow [log1](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630884789986791424),[log2](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630900149062471680),[log3](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630919919547715584)

hope tomorrow we will have more luck 🤞

**Comment by Tai-RedHat** [comment]
> > keep failing on building cluster, will try tomorrow [log1](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630884789986791424),[log2](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630900149062471680),[log3](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1630919919547715584)
> 
> hope tomorrow we will have more luck 🤞

double checked in payload: 4.13.0-0.nightly-2023-03-05-104719, new features implemented

**Pull request merged**